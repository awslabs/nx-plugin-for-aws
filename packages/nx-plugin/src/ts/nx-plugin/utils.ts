/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  joinPathFragments,
  readJson,
  Tree,
  updateJson,
  writeJson,
} from '@nx/devkit';
import { readProjectConfigurationUnqualified } from '../../utils/nx';
import { withVersions } from '../../utils/versions';
import PackageJson from '../../../package.json';

/**
 * Configures a TypeScript project as an Nx Plugin
 */
export const configureTsProjectAsNxPlugin = (
  tree: Tree,
  projectName: string,
) => {
  const project = readProjectConfigurationUnqualified(tree, projectName);

  const tsConfigPath = joinPathFragments(project.root, 'tsconfig.json');
  if (!tree.exists(tsConfigPath)) {
    throw new Error(
      `Selected plugin project ${projectName} is not a TypeScript project`,
    );
  }

  // Nx Plugins must use commonjs as a limitation of nx
  // https://github.com/nrwl/nx/issues/15682
  updateJson(tree, tsConfigPath, (tsConfig) => {
    tsConfig.compilerOptions ??= {};
    tsConfig.compilerOptions.module ??= 'commonjs';
    return tsConfig;
  });

  // Create an empty generators.json if one dosn't exist
  const generatorsJsonPath = joinPathFragments(project.root, 'generators.json');
  if (!tree.exists(generatorsJsonPath)) {
    writeJson(tree, generatorsJsonPath, {
      generators: {},
    });
  }

  // Create a package.json if one doesn't exist, and configure "main" and "generators"
  const pluginPackageJsonPath = joinPathFragments(project.root, 'package.json');
  if (!tree.exists(pluginPackageJsonPath)) {
    writeJson(tree, pluginPackageJsonPath, {
      name: project.name,
    });
  }
  updateJson(tree, pluginPackageJsonPath, (pkg) => {
    pkg.main ??= './src/index.js';
    pkg.generators ??= './generators.json';
    return pkg;
  });

  const rootPackageJson = tree.exists('package.json')
    ? readJson(tree, 'package.json')
    : undefined;
  const isNxPluginForAws = rootPackageJson?.name === '@aws/nx-plugin-source';

  if (!isNxPluginForAws) {
    // Add the required dependencies to the root package.json, and project's package.json
    const deps = {
      ...withVersions(['@nx/devkit']),
      // Include dependency on @aws/nx-plugin
      [PackageJson.name]: `^${PackageJson.version}`,
    };
    addDependenciesToPackageJson(tree, {}, deps);
    addDependenciesToPackageJson(tree, deps, {}, pluginPackageJsonPath);
  }
};
