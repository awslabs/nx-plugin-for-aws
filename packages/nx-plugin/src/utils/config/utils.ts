/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { generateFiles, joinPathFragments, Tree } from '@nx/devkit';
import { AwsNxPluginConfig } from '.';
import * as ts from 'typescript';
import { jsonToAst, replace } from '../ast';
import { factory } from 'typescript';
import { formatFilesInSubtree } from '../format';
import { importTypeScriptModule } from '../js';

export const AWS_NX_PLUGIN_CONFIG_FILE_NAME = 'aws-nx-plugin.config.mts';

/**
 * Ensure that the config file exists
 */
export const ensureAwsNxPluginConfig = async (
  tree: Tree,
): Promise<AwsNxPluginConfig> => {
  if (!tree.exists(AWS_NX_PLUGIN_CONFIG_FILE_NAME)) {
    // Create an empty config file if it doesn't already exist
    generateFiles(tree, joinPathFragments(__dirname, 'files'), '.', {});
  }
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return await readAwsNxPluginConfig(tree)!;
};

/**
 * Read config from the aws nx plugin configuration file
 */
export const readAwsNxPluginConfig = async (
  tree: Tree,
): Promise<AwsNxPluginConfig | undefined> => {
  if (!tree.exists(AWS_NX_PLUGIN_CONFIG_FILE_NAME)) {
    return undefined;
  }
  const configTs = tree.read(AWS_NX_PLUGIN_CONFIG_FILE_NAME, 'utf-8');
  return await importTypeScriptModule(configTs);
};

/**
 * Update the aws nx plugin config file.
 * Undefined top level keys in the config update will be untouched, otherwise config is replaced
 */
export const updateAwsNxPluginConfig = async (
  tree: Tree,
  configUpdate: Partial<AwsNxPluginConfig>,
): Promise<void> => {
  // Replace the default export
  replace(
    tree,
    AWS_NX_PLUGIN_CONFIG_FILE_NAME,
    'ExportAssignment ObjectLiteralExpression',
    (node) => {
      const existingObj = node as ts.ObjectLiteralExpression;

      const existingProps = new Map<string, ts.PropertyAssignment>();
      existingObj.properties.forEach((prop) => {
        if (ts.isPropertyAssignment(prop)) {
          existingProps.set(prop.name.getText(), prop);
        }
      });

      const properties: ts.PropertyAssignment[] = [];

      for (const [key, value] of Object.entries(configUpdate)) {
        properties.push(
          factory.createPropertyAssignment(
            key,
            jsonToAst(value) as ts.Expression,
          ),
        );
      }

      existingObj.properties.forEach((prop) => {
        if (ts.isPropertyAssignment(prop)) {
          const name = prop.name.getText();
          if (configUpdate[name] === undefined) {
            properties.push(prop);
          }
        }
      });

      return factory.createObjectLiteralExpression(properties, true);
    },
  );

  // Format the config nicely after an update
  await formatFilesInSubtree(tree, AWS_NX_PLUGIN_CONFIG_FILE_NAME);
};
