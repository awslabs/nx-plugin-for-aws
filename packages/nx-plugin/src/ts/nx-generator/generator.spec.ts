/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { addProjectConfiguration, Tree, writeJson } from '@nx/devkit';
import {
  tsNxGeneratorGenerator,
  NX_GENERATOR_GENERATOR_INFO,
} from './generator';
import { createTreeUsingTsSolutionSetup } from '../../utils/test';
import { expectHasMetricTags } from '../../utils/metrics.spec';
import { sharedConstructsGenerator } from '../../utils/shared-constructs';
import NxPluginForAwsPackageJson from '../../../package.json';
import NxPluginForAwsProjectJson from '../../../project.json';
import * as fs from 'fs';
import * as path from 'path';

describe('nx-generator generator', () => {
  describe('within @aws/nx-plugin', () => {
    let tree: Tree;

    beforeEach(() => {
      tree = createTreeUsingTsSolutionSetup();

      const pathToRoot = path.resolve(__dirname, '../../../../..');

      // Copy relevant files from this repo into the tree
      [
        'package.json',
        'packages/nx-plugin/project.json',
        'packages/nx-plugin/package.json',
        'packages/nx-plugin/tsconfig.json',
        'docs/astro.config.mjs',
      ].forEach((file) => {
        tree.write(file, fs.readFileSync(path.join(pathToRoot, file)));
      });

      writeJson(tree, 'packages/nx-plugin/generators.json', {
        generators: {
          existing: {
            factory: './some/generator',
            schema: './some/schema.json',
            description: 'existing generator',
            metric: 'g40',
          },
          another: {
            factory: './some/other/generator',
            schema: './some/other/schema.json',
            description: 'other existing generator',
            metric: 'g41',
          },
        },
      });
    });

    it('should work when description is omitted', async () => {
      await tsNxGeneratorGenerator(tree, {
        pluginProject: NxPluginForAwsProjectJson.name,
        name: 'no#description',
      });

      // Check that the files exist
      const generatorDir = 'packages/nx-plugin/src/no-description';
      expect(tree.exists(`${generatorDir}/schema.json`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/schema.d.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.spec.ts`)).toBeTruthy();

      // Check generators.json was updated with empty description
      const generatorsJson = JSON.parse(
        tree.read('packages/nx-plugin/generators.json', 'utf-8'),
      );
      expect(generatorsJson.generators['no#description']).toBeDefined();
      expect(generatorsJson.generators['no#description'].description).toBe(
        'TODO: Add short description of the generator',
      );
    });

    it('should generate an example schema, generator and test', async () => {
      await tsNxGeneratorGenerator(tree, {
        pluginProject: NxPluginForAwsProjectJson.name,
        name: 'foo#bar',
        description: 'Some description',
      });

      // Check that the files exist
      const generatorDir = 'packages/nx-plugin/src/foo-bar';
      expect(tree.exists(`${generatorDir}/schema.json`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/schema.d.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.spec.ts`)).toBeTruthy();

      // Snapshot the files
      expect(tree.read(`${generatorDir}/schema.json`, 'utf-8')).toMatchSnapshot(
        'schema.json',
      );
      expect(tree.read(`${generatorDir}/schema.d.ts`, 'utf-8')).toMatchSnapshot(
        'schema.d.ts',
      );
      expect(
        tree.read(`${generatorDir}/generator.ts`, 'utf-8'),
      ).toMatchSnapshot('generator.ts');
      expect(
        tree.read(`${generatorDir}/generator.spec.ts`, 'utf-8'),
      ).toMatchSnapshot('generator.spec.ts');
    });

    it('should add guide page to docs', async () => {
      await tsNxGeneratorGenerator(tree, {
        pluginProject: NxPluginForAwsProjectJson.name,
        name: 'foo#bar',
        description: 'Some description',
      });

      // Check that the guide page exists
      const guidePath = 'docs/src/content/docs/en/guides/foo-bar.mdx';
      expect(tree.exists(guidePath)).toBeTruthy();

      // Snapshot the guide page
      expect(tree.read(guidePath, 'utf-8')).toMatchSnapshot('guide-page.mdx');

      // Check that the link was added to astro.config.mjs
      const astroConfig = tree.read('docs/astro.config.mjs', 'utf-8');
      expect(astroConfig).toContain("label: 'foo#bar'");
      expect(astroConfig).toContain("link: '/guides/foo-bar'");
    });

    it('should update generators.json', async () => {
      await tsNxGeneratorGenerator(tree, {
        pluginProject: NxPluginForAwsProjectJson.name,
        name: 'foo#bar',
        description: 'Some description',
      });

      // Check generators.json was updated
      const generatorsJson = JSON.parse(
        tree.read('packages/nx-plugin/generators.json', 'utf-8'),
      );
      expect(generatorsJson.generators['foo#bar']).toBeDefined();
      expect(generatorsJson.generators['foo#bar'].factory).toBe(
        './src/foo-bar/generator',
      );
      expect(generatorsJson.generators['foo#bar'].schema).toBe(
        './src/foo-bar/schema.json',
      );
      expect(generatorsJson.generators['foo#bar'].metric).toBe('g42');

      // Snapshot generators.json
      expect(
        tree.read('packages/nx-plugin/generators.json', 'utf-8'),
      ).toMatchSnapshot('generators.json');

      // Check package.json has generators entry
      const packageJson = JSON.parse(
        tree.read('packages/nx-plugin/package.json', 'utf-8'),
      );
      expect(packageJson.generators).toBe('./generators.json');
    });

    it('should update generators.json with out of order metric', async () => {
      writeJson(tree, 'packages/nx-plugin/generators.json', {
        generators: {
          existing: {
            factory: './some/generator',
            schema: './some/schema.json',
            description: 'existing generator',
            metric: 'g51',
          },
          another: {
            factory: './some/other/generator',
            schema: './some/other/schema.json',
            description: 'other existing generator',
            metric: 'g3',
          },
        },
      });

      await tsNxGeneratorGenerator(tree, {
        pluginProject: NxPluginForAwsProjectJson.name,
        name: 'foo#bar',
        description: 'Some description',
      });

      // Check generators.json was updated
      const generatorsJson = JSON.parse(
        tree.read('packages/nx-plugin/generators.json', 'utf-8'),
      );
      expect(generatorsJson.generators['foo#bar'].metric).toBe('g52');
    });

    it('should handle an empty list of existing generators', async () => {
      // Reset generators.json to have an empty generators object
      writeJson(tree, 'packages/nx-plugin/generators.json', {
        generators: {},
      });

      await tsNxGeneratorGenerator(tree, {
        pluginProject: NxPluginForAwsProjectJson.name,
        name: 'empty#test',
        description: 'Some description',
      });

      // Check that the metric is g1
      const generatorsJson = JSON.parse(
        tree.read('packages/nx-plugin/generators.json', 'utf-8'),
      );
      expect(generatorsJson.generators['empty#test'].metric).toBe('g1');
    });

    it('should support a nested directory', async () => {
      await tsNxGeneratorGenerator(tree, {
        pluginProject: NxPluginForAwsProjectJson.name,
        name: 'nested#test',
        directory: 'nested/dir',
        description: 'Some description',
      });

      // Check that the files exist in the nested directory
      const generatorDir = 'packages/nx-plugin/src/nested/dir';
      expect(tree.exists(`${generatorDir}/schema.json`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/schema.d.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.spec.ts`)).toBeTruthy();

      // Check generators.json was updated with correct paths
      const generatorsJson = JSON.parse(
        tree.read('packages/nx-plugin/generators.json', 'utf-8'),
      );
      expect(generatorsJson.generators['nested#test']).toBeDefined();
      expect(generatorsJson.generators['nested#test'].factory).toBe(
        './src/nested/dir/generator',
      );
      expect(generatorsJson.generators['nested#test'].schema).toBe(
        './src/nested/dir/schema.json',
      );
    });

    it('should throw if given a project that is not @aws/nx-plugin', async () => {
      // Add a different project configuration
      addProjectConfiguration(tree, 'other-project', {
        root: 'packages/other-project',
        sourceRoot: 'packages/other-project/src',
      });
      tree.write('packages/other-project/tsconfig.json', '{}');

      // Expect the generator to throw an error
      await expect(
        tsNxGeneratorGenerator(tree, {
          pluginProject: 'other-project',
          name: 'should#fail',
          description: 'Some description',
        }),
      ).rejects.toThrow(
        `Generators should be added to the ${NxPluginForAwsPackageJson.name} project.`,
      );
    });
  });

  describe('within another workspace', () => {
    let tree: Tree;

    beforeEach(() => {
      tree = createTreeUsingTsSolutionSetup();

      writeJson(tree, 'package.json', {
        name: '@myorg/monorepo',
        type: 'module',
      });

      writeJson(tree, 'tools/plugin/tsconfig.json', {});

      addProjectConfiguration(tree, '@test/plugin', {
        root: 'tools/plugin',
        sourceRoot: 'tools/plugin/src',
      });
    });

    it('should throw an error when the project has no tsconfig', async () => {
      // Create a new project without a tsconfig
      addProjectConfiguration(tree, '@test/no-tsconfig', {
        root: 'tools/no-tsconfig',
        sourceRoot: 'tools/no-tsconfig/src',
      });

      // Expect the generator to throw an error
      await expect(
        tsNxGeneratorGenerator(tree, {
          pluginProject: '@test/no-tsconfig',
          name: 'should#fail',
          description: 'Some description',
        }),
      ).rejects.toThrow(
        'Selected plugin project @test/no-tsconfig is not a TypeScript project',
      );
    });

    it('should throw when there is no root package.json', async () => {
      // Remove the root package.json
      tree.delete('package.json');

      await expect(
        tsNxGeneratorGenerator(tree, {
          pluginProject: '@test/plugin',
          name: 'no#root#pkg',
          description: 'Generator with no root package.json',
        }),
      ).rejects.toThrow();
    });

    it('should work when generators.json is empty', async () => {
      // Create an empty generators.json
      writeJson(tree, 'tools/plugin/generators.json', {});

      await tsNxGeneratorGenerator(tree, {
        pluginProject: '@test/plugin',
        name: 'empty#generators',
        description: 'Generator with empty generators.json',
      });

      // Check that the files exist
      const generatorDir = 'tools/plugin/src/empty-generators';
      expect(tree.exists(`${generatorDir}/schema.json`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/schema.d.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.spec.ts`)).toBeTruthy();

      // Check generators.json was updated correctly
      const generatorsJson = JSON.parse(
        tree.read('tools/plugin/generators.json', 'utf-8'),
      );
      expect(generatorsJson.generators).toBeDefined();
      expect(generatorsJson.generators['empty#generators']).toBeDefined();
      expect(generatorsJson.generators['empty#generators'].factory).toBe(
        './src/empty-generators/generator',
      );
      expect(generatorsJson.generators['empty#generators'].schema).toBe(
        './src/empty-generators/schema.json',
      );
      expect(generatorsJson.generators['empty#generators'].description).toBe(
        'Generator with empty generators.json',
      );
    });

    it('should work when description is omitted', async () => {
      await tsNxGeneratorGenerator(tree, {
        pluginProject: '@test/plugin',
        name: 'no#description',
      });

      // Check that the files exist
      const generatorDir = 'tools/plugin/src/no-description';
      expect(tree.exists(`${generatorDir}/schema.json`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/schema.d.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.spec.ts`)).toBeTruthy();

      // Check generators.json was updated with empty description
      const generatorsJson = JSON.parse(
        tree.read('tools/plugin/generators.json', 'utf-8'),
      );
      expect(generatorsJson.generators['no#description']).toBeDefined();
      expect(generatorsJson.generators['no#description'].description).toBe(
        'TODO: Add short description of the generator',
      );
    });

    it('should generate an example schema, generator and test', async () => {
      await tsNxGeneratorGenerator(tree, {
        pluginProject: '@test/plugin',
        name: 'foo#bar',
        description: 'Some description',
      });

      // Check that the files exist
      const generatorDir = 'tools/plugin/src/foo-bar';
      expect(tree.exists(`${generatorDir}/schema.json`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/schema.d.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.spec.ts`)).toBeTruthy();

      // Snapshot the files
      expect(tree.read(`${generatorDir}/schema.json`, 'utf-8')).toMatchSnapshot(
        'local-schema.json',
      );
      expect(tree.read(`${generatorDir}/schema.d.ts`, 'utf-8')).toMatchSnapshot(
        'local-schema.d.ts',
      );
      expect(
        tree.read(`${generatorDir}/generator.ts`, 'utf-8'),
      ).toMatchSnapshot('local-generator.ts');
      expect(
        tree.read(`${generatorDir}/generator.spec.ts`, 'utf-8'),
      ).toMatchSnapshot('local-generator.spec.ts');
    });

    it('should support a nested directory', async () => {
      await tsNxGeneratorGenerator(tree, {
        pluginProject: '@test/plugin',
        name: 'nested#test',
        directory: 'nested/dir',
        description: 'Some description',
      });

      // Check that the files exist in the nested directory
      const generatorDir = 'tools/plugin/src/nested/dir';
      expect(tree.exists(`${generatorDir}/schema.json`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/schema.d.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.spec.ts`)).toBeTruthy();

      // Check generators.json was updated with correct paths
      const generatorsJson = JSON.parse(
        tree.read('tools/plugin/generators.json', 'utf-8'),
      );
      expect(generatorsJson.generators['nested#test']).toBeDefined();
      expect(generatorsJson.generators['nested#test'].factory).toBe(
        './src/nested/dir/generator',
      );
      expect(generatorsJson.generators['nested#test'].schema).toBe(
        './src/nested/dir/schema.json',
      );
    });

    it('should update generators.json', async () => {
      await tsNxGeneratorGenerator(tree, {
        pluginProject: '@test/plugin',
        name: 'foo#bar',
        description: 'Some description',
      });

      // Check generators.json was updated
      const generatorsJson = JSON.parse(
        tree.read('tools/plugin/generators.json', 'utf-8'),
      );
      expect(generatorsJson.generators['foo#bar']).toBeDefined();
      expect(generatorsJson.generators['foo#bar'].factory).toBe(
        './src/foo-bar/generator',
      );
      expect(generatorsJson.generators['foo#bar'].schema).toBe(
        './src/foo-bar/schema.json',
      );
      expect(generatorsJson.generators['foo#bar'].metric).toBeUndefined(); // No metric in non-AWS plugin

      // Snapshot generators.json
      expect(
        tree.read('tools/plugin/generators.json', 'utf-8'),
      ).toMatchSnapshot('local-generators.json');
    });

    it('should create package.json if it does not exist', async () => {
      // Ensure package.json doesn't exist
      expect(tree.exists('tools/plugin/package.json')).toBeFalsy();

      await tsNxGeneratorGenerator(tree, {
        pluginProject: '@test/plugin',
        name: 'create#pkg',
      });

      // Check package.json was created with the correct content
      expect(tree.exists('tools/plugin/package.json')).toBeTruthy();
      const packageJson = JSON.parse(
        tree.read('tools/plugin/package.json', 'utf-8'),
      );
      expect(packageJson.name).toBe('@test/plugin');
      expect(packageJson.generators).toBe('./generators.json');
    });

    it('should update package.json if it exists but does not have generators entry', async () => {
      // Create package.json without generators entry
      writeJson(tree, 'tools/plugin/package.json', {
        name: '@test/plugin',
        version: '1.0.0',
      });

      await tsNxGeneratorGenerator(tree, {
        pluginProject: '@test/plugin',
        name: 'update#pkg',
      });

      // Check package.json was updated with generators entry
      const packageJson = JSON.parse(
        tree.read('tools/plugin/package.json', 'utf-8'),
      );
      expect(packageJson.name).toBe('@test/plugin');
      expect(packageJson.version).toBe('1.0.0');
      expect(packageJson.generators).toBe('./generators.json');
    });

    it('should not modify existing generators entry in package.json', async () => {
      // Create package.json with custom generators entry
      writeJson(tree, 'tools/plugin/package.json', {
        name: '@test/plugin',
        version: '1.0.0',
        generators: './custom-generators.json',
      });

      await tsNxGeneratorGenerator(tree, {
        pluginProject: '@test/plugin',
        name: 'preserve#generators',
      });

      // Check package.json still has the original generators entry
      const packageJson = JSON.parse(
        tree.read('tools/plugin/package.json', 'utf-8'),
      );
      expect(packageJson.generators).toBe('./custom-generators.json');
    });

    it('should generate an example template file', async () => {
      await tsNxGeneratorGenerator(tree, {
        pluginProject: '@test/plugin',
        name: 'template#test',
        description: 'Some description',
      });

      // Check that the template file exists
      const templatePath =
        'tools/plugin/src/template-test/files/hello.ts.template';
      expect(tree.exists(templatePath)).toBeTruthy();

      // Snapshot the template file
      expect(tree.read(templatePath, 'utf-8')).toMatchSnapshot(
        'hello.ts.template',
      );
    });

    it('should set module to commonjs in tsconfig.json', async () => {
      // Set up a tsconfig.json without module set
      writeJson(tree, 'tools/plugin/tsconfig.json', {});

      await tsNxGeneratorGenerator(tree, {
        pluginProject: '@test/plugin',
        name: 'module#test',
        description: 'Test module setting',
      });

      // Check that module is set to commonjs in tsconfig.json
      const tsConfig = JSON.parse(
        tree.read('tools/plugin/tsconfig.json', 'utf-8'),
      );
      expect(tsConfig.compilerOptions.module).toBe('commonjs');
    });

    it('should work when there is no sourceRoot', async () => {
      // Create a project configuration without sourceRoot
      addProjectConfiguration(tree, '@test/no-source-root', {
        root: 'tools/no-source-root',
      });

      // Create tsconfig.json for the project
      writeJson(tree, 'tools/no-source-root/tsconfig.json', {});

      await tsNxGeneratorGenerator(tree, {
        pluginProject: '@test/no-source-root',
        name: 'no#source#root',
        description: 'Generator in project without sourceRoot',
      });

      // Check that the files exist in the default src directory
      const generatorDir = 'tools/no-source-root/src/no-source-root';
      expect(tree.exists(`${generatorDir}/schema.json`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/schema.d.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.ts`)).toBeTruthy();
      expect(tree.exists(`${generatorDir}/generator.spec.ts`)).toBeTruthy();
    });

    it('should add an export declaration to index.ts', async () => {
      // Create an index.ts file in the source root
      tree.write('tools/plugin/src/index.ts', '// This is the index file');

      await tsNxGeneratorGenerator(tree, {
        pluginProject: '@test/plugin',
        name: 'export#test',
        description: 'Generator with export test',
      });

      // Check that the export was added to index.ts
      const indexContent = tree.read('tools/plugin/src/index.ts', 'utf-8');
      expect(indexContent).toContain(
        "export * from './export-test/generator';",
      );
    });

    it('should add generator metric to app.ts', async () => {
      await sharedConstructsGenerator(tree);

      await tsNxGeneratorGenerator(tree, {
        pluginProject: '@test/plugin',
        name: 'foo#bar-baz',
      });

      expectHasMetricTags(tree, NX_GENERATOR_GENERATOR_INFO.metric);
    });
  });
});
