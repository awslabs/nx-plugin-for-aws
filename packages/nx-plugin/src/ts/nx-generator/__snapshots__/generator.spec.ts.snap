// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`nx-generator generator > within @aws/nx-plugin > should add guide page to docs > guide-page.mdx 1`] = `
"---
title: foo#bar
description: Some description
---

import { FileTree } from '@astrojs/starlight/components';
import RunGenerator from '@components/run-generator.astro';
import GeneratorParameters from '@components/generator-parameters.astro';

TODO: Add an overview of your generator here.

## Usage

### Generate a foo#bar

You can generate a foo#bar in two ways:

<RunGenerator generator="foo#bar" />

### Options

<GeneratorParameters generator="foo#bar" />

## Generator Output

The generator will create the following project files:

<FileTree>
  - todo.txt Update this section with what your generator generates!
</FileTree>

## TODO: More Docs!

Add more documentation detailing things like:

- How to work with the project/code you've generated
- What new project.json targets have been added and how to run them (using the NxCommands component)
- Best practices when working with your generated project/code

See the existing guides for concrete examples.
"
`;

exports[`nx-generator generator > within @aws/nx-plugin > should generate an example schema, generator and test > generator.spec.ts 1`] = `
"import { Tree } from '@nx/devkit';
import { fooBarGenerator, FOO_BAR_GENERATOR_INFO } from './generator';
import { createTreeUsingTsSolutionSetup } from '../utils/test';
import { expectHasMetricTags } from '../utils/metrics.spec';
import { sharedConstructsGenerator } from '../utils/shared-constructs';

describe('foo#bar generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
  });

  it('should run successfully', async () => {
    await fooBarGenerator(tree, { exampleOption: 'example' });

    // TODO: check the tree is updated as expected
  });

  it('should add generator metric to app.ts', async () => {
    await sharedConstructsGenerator(tree);

    await fooBarGenerator(tree, { exampleOption: 'example' });

    expectHasMetricTags(tree, FOO_BAR_GENERATOR_INFO.metric);
  });
});
"
`;

exports[`nx-generator generator > within @aws/nx-plugin > should generate an example schema, generator and test > generator.ts 1`] = `
"import { GeneratorCallback, Tree, installPackagesTask } from '@nx/devkit';
import { FooBarGeneratorSchema } from './schema';
import { NxGeneratorInfo, getGeneratorInfo } from '../utils/nx';
import { addGeneratorMetricsIfApplicable } from '../utils/metrics';
import { formatFilesInSubtree } from '../utils/format';

export const FOO_BAR_GENERATOR_INFO: NxGeneratorInfo =
  getGeneratorInfo(__filename);

export const fooBarGenerator = async (
  tree: Tree,
  options: FooBarGeneratorSchema,
): Promise<GeneratorCallback> => {
  // TODO: implement your generator here

  await addGeneratorMetricsIfApplicable(tree, [FOO_BAR_GENERATOR_INFO]);

  await formatFilesInSubtree(tree);
  return () => {
    installPackagesTask(tree);
  };
};

export default fooBarGenerator;
"
`;

exports[`nx-generator generator > within @aws/nx-plugin > should generate an example schema, generator and test > schema.d.ts 1`] = `
"/**
 * TypeScript types for options defined in schema.json
 * Update this to match schema.json if you make changes.
 */
export interface FooBarGeneratorSchema {
  // Replace with your options
  exampleOption: string;
}
"
`;

exports[`nx-generator generator > within @aws/nx-plugin > should generate an example schema, generator and test > schema.json 1`] = `
"{
  "$schema": "https://json-schema.org/schema",
  "$id": "foo#bar",
  "title": "foo#bar",
  "description": "<Describe your generator here!>",
  "type": "object",
  "properties": {
    "exampleOption": {
      "type": "string",
      "description": "An example option. Refer to https://nx.dev/extending-nx/recipes/generator-options for more info",
      "x-priority": "important",
      "x-prompt": "This is an example option!"
    }
  },
  "required": ["exampleOption"]
}
"
`;

exports[`nx-generator generator > within @aws/nx-plugin > should update generators.json > generators.json 1`] = `
"{
  "generators": {
    "existing": {
      "factory": "./some/generator",
      "schema": "./some/schema.json",
      "description": "existing generator",
      "metric": "g40"
    },
    "another": {
      "factory": "./some/other/generator",
      "schema": "./some/other/schema.json",
      "description": "other existing generator",
      "metric": "g41"
    },
    "foo#bar": {
      "factory": "./src/foo-bar/generator",
      "schema": "./src/foo-bar/schema.json",
      "description": "Some description",
      "metric": "g42"
    }
  }
}
"
`;

exports[`nx-generator generator > within another workspace > should generate an example schema, generator and test > local-generator.spec.ts 1`] = `
"import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { Tree } from '@nx/devkit';
import { fooBarGenerator } from './generator';

describe('foo#bar generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  it('should run successfully', async () => {
    await fooBarGenerator(tree, { exampleOption: 'example' });

    // TODO: check the tree is updated as expected
  });
});
"
`;

exports[`nx-generator generator > within another workspace > should generate an example schema, generator and test > local-generator.ts 1`] = `
"import {
  GeneratorCallback,
  Tree,
  generateFiles,
  joinPathFragments,
} from '@nx/devkit';
import { FooBarGeneratorSchema } from './schema';

export const fooBarGenerator = async (
  tree: Tree,
  options: FooBarGeneratorSchema,
): Promise<GeneratorCallback | void> => {
  // TODO: implement your generator here, for example:
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    'target/dir',
    options,
  );
};

export default fooBarGenerator;
"
`;

exports[`nx-generator generator > within another workspace > should generate an example schema, generator and test > local-schema.d.ts 1`] = `
"/**
 * TypeScript types for options defined in schema.json
 * Update this to match schema.json if you make changes.
 */
export interface FooBarGeneratorSchema {
  // Replace with your options
  exampleOption: string;
}
"
`;

exports[`nx-generator generator > within another workspace > should generate an example schema, generator and test > local-schema.json 1`] = `
"{
  "$schema": "https://json-schema.org/schema",
  "$id": "foo#bar",
  "title": "foo#bar",
  "description": "<Describe your generator here!>",
  "type": "object",
  "properties": {
    "exampleOption": {
      "type": "string",
      "description": "An example option. Refer to https://nx.dev/extending-nx/recipes/generator-options for more info",
      "x-priority": "important",
      "x-prompt": "This is an example option!"
    }
  },
  "required": ["exampleOption"]
}
"
`;

exports[`nx-generator generator > within another workspace > should generate an example template file > hello.ts.template 1`] = `
"
export const hello = () => {
  return '<%- exampleOption %>';
};
"
`;

exports[`nx-generator generator > within another workspace > should update generators.json > local-generators.json 1`] = `
"{
  "generators": {
    "foo#bar": {
      "factory": "./src/foo-bar/generator",
      "schema": "./src/foo-bar/schema.json",
      "description": "Some description"
    }
  }
}
"
`;
